[SU-250] Support Azure workspaces in new file browser

nawatts opened the pull request:
This enables the new file browser for Azure workspaces by implementing a version of the FileBrowserProvider interface for Azure Blob Storage. Files are stored in the workspace's storage container. The storage container URL can be found in the Cloud Information panel on the workspace dashboard.

Enable the new file browser from the feature preview page (/#feature-preview) and open it from the right sidebar within a workspace.

nawatts committed:
Add file browser provider for Azure Blob Storage

nawatts committed:
Enable file browser for Azure workspaces

cahrens commented:
I'm struggling to get an environment in which I can try this out (CORS). But where is the "Workspace bucket" link supposed to take me? It is styled as a link, but it's a button?

<img width="773" alt="image" src="https://user-images.githubusercontent.com/484484/207640649-5143db7e-8296-4bd5-bd2b-baf80d89c73e.png">

nawatts commented:
I shared https://bvdp-saturn-dev.appspot.com/#workspaces/azure-dev-2022-12-02/my-first-azure-workspace with you, which has the necessary CORS settings configured. The links/buttons above the files list show which directory files are listed for and serve as breadcrumbs.

nawatts committed:
Remove SAS tokens from file URLs

Matches behavior of current file browser with GCS objects, where the
file links in the browser are unauthenticated URLs.

cahrens commented on src/pages/workspaces/workspace/Files.ts:
```diff
@@ -17,11 +18,17 @@ export const Files = _.flow(
-    () => GCSFileBrowserProvider({ bucket: bucketName, project: googleProject }),
-    [bucketName, googleProject]
+    () => {
+      if (workspace.azureContext) {
```
Should also be able to use `workspace.cloudPlatform` now.

nawatts replied:
Updated this to use `isAzureWorkspace` from workspace-utils. That function does currently use `azureContext`,b ut we can change that later. https://github.com/DataBiosphere/terra-ui/pull/3580/commits/422f53f2ea59dfe8d86f3e4efdd75e5992912aab.

cahrens commented on src/libs/ajax/file-browser-providers/AzureBlobStorageFileBrowserProvider.ts:
```diff
@@ -0,0 +1,185 @@
+          const name = blob.getElementsByTagName('Name').item(0)!.textContent!
+
+          const blobProperties = blob.getElementsByTagName('Properties').item(0)
+          const creationTime = blobProperties?.getElementsByTagName('Creation-Time').item(0)!.textContent!
```
Do the exclamation marks mean that an error will be thrown if the variable is undefined?

nawatts replied:
Yes. 

`.item(0)` [returns null if the collection is empty](https://developer.mozilla.org/en-US/docs/Web/API/HTMLCollection/item#return_value) / if no elements with that tag were found. However, we assume that the [API response will include these tags](https://learn.microsoft.com/en-us/rest/api/storageservices/list-blobs?tabs=azure-ad), so the exclamation mark ([non-null assertion operator](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-0.html#non-null-assertion-operator)) changes the return type of `item(0)` from `Node | null` to `Node`.

Similar with `textContent`: [Node.textContent](https://developer.mozilla.org/en-US/docs/Web/API/Node/textContent) is typed `string | null`. We know this is not a case where `textContent` would return null, so `textContent!` is used to change the type to `string`.

Just noticed that `blobProperties` has a `?` instead of a `!`. Will change that.

nawatts committed:
Assert blobProperties is non-null

nawatts committed:
Use isAzureWorkspace to check if workspace is an Azure workspace
